pkg/controller/service/servicecontroller_test.go

```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		testName        string
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
+	}{
+		{
+			testName: "If an invalid service is deleted",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				expectedError := "Service external-balancer not in cache even though the watcher thought it was. Ignoring the deletion."
+				if srvErr == nil || srvErr.Error() != expectedError {
+					//cannot be nil or Wrong error message
+					return fmt.Errorf("Expected=%v Obtained=%v", expectedError, srvErr)
+				}
+
+				if retryDuration != doNotRetry {
+					//Retry duration should match
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+
+				return nil
+			},
+		},
+		{
+			testName: "If cloudprovided failed to delete the service",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+				cloud.Err = fmt.Errorf("Error Deleting the Loadbalancer")
+
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				expectedError := "Error Deleting the Loadbalancer"
+
+				if srvErr == nil || srvErr.Error() != expectedError {
+					return fmt.Errorf("Expected=%v Obtained=%v", expectedError, srvErr)
+				}
+
+				if retryDuration != minRetryDelay {
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", minRetryDelay, retryDuration)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "If delete was successful",
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				if srvErr != nil {
+					return fmt.Errorf("Expected=nil Obtained=%v", srvErr)
+				}
+
+				if retryDuration != doNotRetry {
+					//Retry duration should match
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		obtainedErr, retryDuration := controller.processServiceDeletion("external-balancer")
+		if err := tst.checkExpectedFn(obtainedErr, retryDuration); err != nil {
+			t.Errorf("%v processServiceDeletion() %v", tst.testName, err)
+		}
+	}
+
+}
+
+// Add unit testing for needsUpdate functions
+
+func TestDoesExternalLoadBalancerNeedsUpdate(t *testing.T) {
+
+	var oldSrv, newSrv *v1.Service
+
+	testCases := []struct {
+		testName       string //Name of the test case
+		updateFn       func() //Function to update the old and new service varuables
```

Comment:
The way you are using updateFn here is not intuitive. Usually I would expect oldService and newService as fields in the testcase struct. Though I could see making this a function make it simpler for tweaking both old and new services. We could keep it for now, but it will be great if you have better idea :)